### ISSUE TITLE
Confusing difference between types and interfaces when intersected with string indexes

### ISSUE THREAD

**Issue by ekilah** [Microsoft/TypeScript#24970]
**TypeScript Version:**  2.8.3

**Search Terms:** intersection types and interfaces seem to behave differently. Inconsistent 'property 'x' of type 'y' is not assignable to string index type

**Code**

```ts
interface IJSONObject {
  [key: string]: string | number | boolean | IJSONObject
}

// we want to enforce that some objects have an `id` field
// with a `type`. 
type TIdElement = {
  id: string
}

// here's an object with an `id` and a `day`.
// we've enforced the `id` member with a `type`.
type WithType = TIdElement & {
    day: Date,
}

// we could use an interface to enforce the same idea that
// things have an `id`
interface IIdElement {
  id: string
}

// here's another object with an `id` and a `day`,
// this time enforced with an `interface` instead.
type WithInterface = IIdElement & {
    day: Date,
}

// only the version where `id` was added via an intersection
// with an `interface` causes a type error.
//
// expected behavior: a type error should happen with both,
// because `Date` is not in `IJSONObject`'s index signature
interface ITest extends IJSONObject {
    baz: WithType // no error, why?
    foo: WithInterface    // error, as expected
}

```

**Expected behavior:**
a type error should happen with both,
because `Date` is not in `IJSONObject`'s index signature

**Actual behavior:**
`WithType` doesn't cause an error, even though `Date` shouldn't be allowed.

**Playground Link:** [link](https://www.typescriptlang.org/play/#src=interface%20IJSONObject%20%7B%0D%0A%20%20%5Bkey%3A%20string%5D%3A%20string%20%7C%20number%20%7C%20boolean%20%7C%20IJSONObject%0D%0A%7D%0D%0A%0D%0A%2F%2F%20we%20want%20to%20enforce%20that%20some%20objects%20have%20an%20%60id%60%20field%0D%0A%2F%2F%20with%20a%20%60type%60.%20%0D%0Atype%20TIdElement%20%3D%20%7B%0D%0A%20%20id%3A%20string%0D%0A%7D%0D%0A%0D%0A%2F%2F%20here's%20an%20object%20with%20an%20%60id%60%20and%20a%20%60day%60.%0D%0A%2F%2F%20we've%20enforced%20the%20%60id%60%20member%20with%20a%20%60type%60.%0D%0Atype%20WithType%20%3D%20TIdElement%20%26%20%7B%0D%0A%20%20%20%20day%3A%20Date%2C%0D%0A%7D%0D%0A%0D%0A%2F%2F%20we%20could%20use%20an%20interface%20to%20enforce%20the%20same%20idea%20that%0D%0A%2F%2F%20things%20have%20an%20%60id%60%0D%0Ainterface%20IIdElement%20%7B%0D%0A%20%20id%3A%20string%0D%0A%7D%0D%0A%0D%0A%2F%2F%20here's%20another%20object%20with%20an%20%60id%60%20and%20a%20%60day%60%2C%0D%0A%2F%2F%20this%20time%20enforced%20with%20an%20%60interface%60%20instead.%0D%0Atype%20WithInterface%20%3D%20IIdElement%20%26%20%7B%0D%0A%20%20%20%20day%3A%20Date%2C%0D%0A%7D%0D%0A%0D%0A%2F%2F%20only%20the%20version%20where%20%60id%60%20was%20added%20via%20an%20intersection%0D%0A%2F%2F%20with%20an%20%60interface%60%20causes%20a%20type%20error.%0D%0A%2F%2F%0D%0A%2F%2F%20expected%20behavior%3A%20a%20type%20error%20should%20happen%20with%20both%2C%0D%0A%2F%2F%20because%20%60Date%60%20is%20not%20in%20%60IJSONObject%60's%20index%20signature%0D%0Ainterface%20ITest%20extends%20IJSONObject%20%7B%0D%0A%20%20%20%20baz%3A%20WithType%0D%0A%20%20%20%20foo%3A%20WithInterface%20%20%20%20%0D%0A%7D%0D%0A)

**Related Issues:** possibly https://github.com/Microsoft/TypeScript/issues/18075 but I'm not really sure.

**Comment by RyanCavanaugh** [comment]
Intersection is a higher-order process that can occur during e.g. generic instantiation; intentionally this process can never create errors.

`extends` happens immediately and we do check that you fulfill the base type's contract.

**Comment by ekilah** [comment]
@RyanCavanaugh I'm not sure I completely understand you. Both `baz` and `foo` here are added via `extends`, and both `WithType` and `WithInterface` are constructed with intersections (`&`).

**Comment by RyanCavanaugh** [comment]
One of the type rules is that `X` is assignable to `T & U` if `X` is assignable to `T` *or* `X` is assignable to `U`. `{id: string}`, one of the constituents of `WithType`, is assignable to `IJSonObject`.

**Comment by MrDesjardins** [comment]
I am confused like @ekilah on this one. The interface in that situation has one of the constituents assignable to `IJSONObject`, why is it rejected? Your (Ryan) first reply is also confusing since both (type and interface) are using the intersection, and both are a member of a class that extends an interface.

**Comment by RyanCavanaugh** [comment]
The spec defines the rules: https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md

There is not a single place that explains all emergent behavior. Sometimes we define that X causes Y, and B causes C, and P causes Q, and M causes N, then someone notices that X + B - P + Q isn't identical to X + Q - P + M even though M and Q are supposed to be pretty similar - it isn't so much that we specifically designed the X + Q - P + M rule and have written down a convincing explanation of why that should be the case, but rather than the initial five propositions are all individually sensible and cannot be changed without making a lot of other things worse.

**Comment by ford04** [comment]
I also have been baffled by this distinction.  To give another explanation on @MrDesjardins and @ekilah questions: My guess is that type literals work in this case, because they get assigned an[ implicit index signature](https://github.com/microsoft/TypeScript/issues/15300), whereas interfaces index signatures have to be declared explicitly. Test case:
```
interface A { [K: string]: string }
interface B { foo: string }
type O = { foo: string }

type BExtendsA = B extends A ? true : false // false
type OExtendsA = O extends A ? true : false // true
```
So, `WithType` compiles, as on of the intersection parts `TIdElement` gets an implicit index signature and is then assignable to `IJSONObject`. And both parts of `WithInterface` aren't sub types of `IJSONObject`.
```
type T1 = { foo: WithInterface } extends IJSONObject ? true : false // false
type T2 = { foo: WithType } extends IJSONObject ? true : false // true
// true, TIdElement has implicit index signature
type T3 = TIdElement extends IJSONObject ? true : false 
// false, no index sign. for IIdElement 
type T4 = IIdElement extends IJSONObject ? true : false 
```

**Comment by ekilah** [comment]
@ford04 that's an interesting angle, and I really appreciate you sharing it. Your explanation goes above and beyond any of @RyanCavanaugh 's three "attempts" to half-answer this questin or provide some insight after already closing it as if it were obvious. I haven't seen mention of this "implicit index signature" before, but that issue you linked seems to describe a similar issue. Unfortunately it seems like people there have mostly been left to figure it out themselves as well.

For what it's worth, to this day I've never found a good resource that actually describes how types and interfaces differ internally or how we, as external clients of the language, should use them differently, and after using TypeScript for 1.5 years professionally now I think that is simply ridiculous. Here we are, trying to use the language as intended, and when we run across weird behaviors that don't make a lot of sense, the answer is not much more than a "working as intended" label.

@RyanCavanaugh I invite you and your team back to read this thread all the way through to see if you can give us a little better of an answer, all this time later. Individually, none of your answers above make much sense, and together, they sound fairly incoherent. With all the advances of the type system lately, it surprises me that these types of questions are still left mostly unanswered.
